models/Element/Service.php

```diff
@@ -718,31 +718,81 @@ public static function gridElementData(ElementInterface $element)
      */
     public static function findForbiddenPaths($type, $user)
     {
+        $db = Db::get();
+
         if ($user->isAdmin()) {
-            return [];
+            return array('forbidden' => [], 'allowed' => []);
         }
 
-        // get workspaces
-        $workspaces = $user->{'getWorkspaces' . ucfirst($type)}();
-        foreach ($user->getRoles() as $roleId) {
-            $role = Model\User\Role::getById($roleId);
-            $workspaces = array_merge($workspaces, $role->{'getWorkspaces' . ucfirst($type)}());
+        $currentUserId = $user->getId();
+        $userIds = $user->getRoles();
+        $userIds[] = $currentUserId;
+
+        $userWorkspacesSql = '
+            SELECT cpath, cid, list
+            FROM users_workspaces_'.$type.'
+            WHERE userId = '.$currentUserId.'
+        ';
+        $userWorkspaces = $db->fetchAll($userWorkspacesSql);
+
+        //this collects the array that are on user-level, which have top priority
+        $userCid = [];
+        foreach ($userWorkspaces as $userWorkspace){
+            $userCid[] = $userWorkspace['cid'];
         }
 
+        $roleWorkspacesSql = '
+            SELECT
+                cpath, userid, max(list) as list
+            FROM users_workspaces_'.$type.'
+            WHERE userId IN (' . implode(',', $userIds) . ') AND cid NOT IN ('.implode(',', $userCid).')
+            GROUP BY cpath
+        ';
+        $roleWorkspaces = $db->fetchAll($roleWorkspacesSql);
+
+        $allWorkspaces = array_merge($userWorkspaces, $roleWorkspaces);
+
+        $uniquePaths = [];
+        foreach ($allWorkspaces as $workspace){
+            $uniquePaths[$workspace['cpath']] = $workspace['list'];
```

Comment:
doesn't here the role permission potentially overwrite the user permission with the same path?

Reply:
following issue needs to be fixed after this is merged: https://github.com/pimcore/advanced-object-search/issues/149

Reply:
> doesn't here the role permission potentially overwrite the user permission with the same path?

No, it shouldn't, the role permissions query have this condition
`cid NOT IN ('.implode(',', $userCid).')`, which translates that the cpath would be excluded, if found in the first query (which gets the user level)